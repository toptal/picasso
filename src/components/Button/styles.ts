import { Theme, createStyles } from '@material-ui/core/styles'

import { lighten, darken, alpha } from '../styles'

const ICON_SPACING = '0.4em'

const primary = (mainColor: string, secondaryColor: string) => ({
  border: 'none',
  color: secondaryColor,
  backgroundColor: mainColor,

  '&:hover, &$hovered': {
    backgroundColor: darken(mainColor, 0.2)
  },

  '&:active, &$active': {
    backgroundColor: darken(mainColor, 0.2)
  }
})

const secondary = (mainColor: string, secondaryColor: string) => ({
  color: mainColor,
  backgroundColor: secondaryColor,

  '&:hover, &$hovered': {
    backgroundColor: lighten(mainColor, 0.84),
    borderColor: mainColor
  },

  '&:active, &$active': {
    backgroundColor: lighten(mainColor, 0.84),
    borderColor: mainColor
  }
})

export default ({ palette, spacing, transitions, typography }: Theme) =>
  createStyles({
    root: {
      position: 'relative',
      textTransform: 'none',
      borderRadius: spacing.borderRadius,
      border: `solid ${spacing.borderWidth} ${palette.grey[100]}`,
      fontSize: '1rem',
      transition: `all ${transitions.duration.short}ms ${
        transitions.easing.easeOut
      }`,
      transitionProperty: 'border, color, background',

      '&:focus, &$focused': {
        textDecoration: 'underline',

        '&:active, &$active, &:hover': {
          textDecoration: 'none'
        }
      },

      '&+&': {
        marginLeft: '0.5em'
      }
    },
    content: {
      lineHeight: '1.5em',
      fontWeight: typography.fontWeights.semibold
    },
    loader: {
      position: 'absolute',
      left: '50%',
      top: '50%',
      transform: 'translate(-50%, -50%)'
    },

    // sizes
    small: {
      height: '1.5em',
      padding: '0 0.75em',

      '& $content': {
        fontSize: typography.buttons.fontSizeSmall
      },

      '&$circular': {
        width: '1.5em'
      }
    },
    medium: {
      height: '2.25em',
      padding: '0 1em',

      '& $content': {
        fontSize: typography.buttons.fontSizeMedium
      },

      '&$circular': {
        width: '2.25em'
      }
    },
    large: {
      height: '3em',
      padding: '0 3.625em',

      '& $content': {
        fontSize: typography.buttons.fontSizeLarge,
        fontWeight: typography.fontWeights.semibold
      },

      '&$circular': {
        width: '3em'
      }
    },

    // variants
    primaryBlue: primary(palette.primary.main, palette.common.white),
    secondaryBlue: secondary(palette.primary.main, palette.common.white),
    primaryRed: primary(palette.error.main, palette.common.white),
    secondaryRed: secondary(palette.error.main, palette.common.white),
    primaryGreen: primary(palette.success.main, palette.common.white),
    secondaryWhite: {
      ...secondary(palette.common.white, palette.common.white),
      backgroundColor: 'transparent',
      border: `solid ${spacing.borderWidth} rgba(255, 255, 255, 0.32)`,

      '&:hover, &$hovered': {
        backgroundColor: alpha(palette.common.white, 0.16),
        borderColor: palette.common.white
      },

      '&:active, &$active': {
        backgroundColor: alpha(palette.common.white, 0.16),
        borderColor: palette.common.white
      }
    },
    flat: {
      ...secondary(palette.common.black, palette.common.white),
      border: 'none'
    },
    primaryDisabled: primary(palette.grey[100], palette.common.white),
    secondaryDisabled: secondary(palette.grey[100], palette.common.white),
    flatDisabled: {
      ...secondary(palette.grey[100], palette.common.white),
      border: 'none'
    },

    // Other props
    fullWidth: {
      width: '100%'
    },
    hovered: {},
    focused: {},
    active: {},
    circular: {
      borderRadius: '50%',
      padding: 0
    },

    // Child elements
    icon: {
      fontSize: '1.2em !important'
    },
    iconLeft: {
      marginRight: ICON_SPACING
    },
    iconRight: {
      marginLeft: ICON_SPACING
    },
    hidden: {
      opacity: 0
    }
  })
