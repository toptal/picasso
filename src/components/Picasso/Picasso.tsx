import {
  createMuiTheme,
  MuiThemeProvider,
  withStyles
} from '@material-ui/core/styles'
import React, { FunctionComponent, ReactNode } from 'react'

import CssBaseline from '../CssBaseline'
import {
  palette,
  layout,
  transitions,
  typography,
  spacing,
  breakpoints,
  screens,
  shadows
} from './config'
import FontsLoader from './FontsLoader'
import Provider from './PicassoProvider'
import globalStyles from './styles'
import { JssProps } from './types'

const picasso = {
  palette,
  layout,
  transitions,
  spacing,
  breakpoints,
  screens,
  shadows,
  typography,
  props: {
    MuiButtonBase: {
      disableRipple: true
    },
    MuiList: {
      disablePadding: true
    },
    MuiPaper: {
      square: true
    },
    MuiOutlinedInput: {
      notched: false
    }
  }
}

const PicassoProvider = new Provider(createMuiTheme(picasso))

interface PicassoGlobalStylesProviderProps extends JssProps {
  children?: ReactNode
}

const PicassoGlobalStylesProvider = withStyles(globalStyles, {
  name: 'Picasso'
})((props: PicassoGlobalStylesProviderProps) => {
  const { classes, children } = props

  return <div className={classes.root}>{children}</div>
})

interface PicassoProps {
  children?: ReactNode
  /** Whether to load fonts file to the page */
  loadFonts?: boolean
  /** Whether to apply Picasso CSS reset */
  reset?: boolean
}

const Picasso: FunctionComponent<PicassoProps> = ({
  loadFonts,
  reset,
  children
}) => (
  <MuiThemeProvider theme={PicassoProvider.theme}>
    {loadFonts && <FontsLoader />}
    {reset && <CssBaseline />}
    <PicassoGlobalStylesProvider>{children}</PicassoGlobalStylesProvider>
  </MuiThemeProvider>
)

Picasso.defaultProps = {
  loadFonts: true,
  reset: true
}

export { PicassoProvider }
export default Picasso
