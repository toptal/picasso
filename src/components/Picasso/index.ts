export {
  default,
  PicassoProvider,
  usePicassoRoot,
  usePageHeader
} from './Picasso'

export {
  BaseProps,
  JssProps,
  StandardProps,
  SizeType,
  OmitInternalProps,
  SpacingType,
  spacingToEm,
  PicassoComponent,
  PicassoComponentWithRef,
  SpacingEnum,
  ButtonOrAnchorProps,
  CompoundedComponentWithRef,
  OverridableComponent,
  ColorType
} from './types'

export { useScreenSize, isScreenSize, useBreakpoint } from './config'
