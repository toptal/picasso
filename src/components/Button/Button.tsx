import React, { ReactNode, ReactElement, MouseEvent, forwardRef } from 'react'
import cx from 'classnames'
import { withStyles } from '@material-ui/core/styles'
import ButtonBase from '@material-ui/core/ButtonBase'

import Loader from '../Loader'
import Container from '../Container'
import Group from '../ButtonGroup'
import kebabToCamelCase from '../utils/kebab-to-camel-case'
import {
  StandardProps,
  PicassoComponentWithRef,
  SizeType,
  ButtonOrAnchorProps,
  CompoundedComponentWithRef
} from '../Picasso'
import styles from './styles'

type VariantType =
  | 'primary-blue'
  | 'secondary-blue'
  | 'primary-red'
  | 'secondary-red'
  | 'primary-green'
  | 'flat'
  | 'flat-white'
  | 'secondary-white'

type IconPositionType = 'left' | 'right'

export interface Props extends StandardProps, ButtonOrAnchorProps {
  /** Show button in the active state (left mouse button down) */
  active?: boolean
  /** Disables button */
  disabled?: boolean
  /** Content of Button component */
  children?: ReactNode
  // TODO: should it be exposed?
  focused?: boolean
  /** Take the full width of a container */
  fullWidth?: boolean
  /** Set hovered style for the button */
  hovered?: boolean
  /** Add an `<Icon />` along Button's children */
  icon?: ReactElement
  /** Icon can be positioned on the left or right */
  iconPosition?: IconPositionType
  /** A button can show a loading indicator */
  loading?: boolean
  /** Callback invoked when component is clicked */
  onClick?: (event: MouseEvent<HTMLButtonElement & HTMLAnchorElement>) => void
  /** A button can have different sizes */
  size?: SizeType<'small' | 'medium' | 'large'>
  /** The variant to use */
  variant?: VariantType
  /** HTML Value of Button component */
  value?: string | number
  /** Circular style of Button component */
  circular?: boolean
  /** HTML title of Button component */
  title?: string
  /** HTML type of Button component **/
  type?: 'button' | 'reset' | 'submit'
}

interface StaticProps {
  Group: typeof Group
}

const getVariantType = (variant: VariantType) => {
  const [type] = variant!.split('-')

  return type
}

export const Button = forwardRef<HTMLButtonElement, Props>(function Button(
  {
    icon,
    iconPosition,
    loading,
    children,
    classes,
    className,
    style,
    fullWidth,
    variant,
    size,
    focused,
    hovered,
    disabled,
    active,
    onClick,
    circular,
    title,
    value,
    type,
    ...rest
  },
  ref
) {
  const {
    icon: iconClass,
    iconLeft: iconLeftClass,
    iconRight: iconRightClass,
    iconSmall: iconSmallClass,
    root: rootClass,
    hidden: hiddenClass,
    loader: loaderClass,
    content: contentClass
  } = classes

  let finalChildren = [children]

  if (icon) {
    const iconComponent = React.cloneElement(icon, {
      className: cx(iconClass, icon.props.className, {
        [iconLeftClass]: children && iconPosition === 'left',
        [iconRightClass]: children && iconPosition === 'right',
        [iconSmallClass]: size === 'small'
      }),
      key: 'button-icon',
      base: size === 'small' ? '12' : undefined
    })

    if (iconPosition === 'left') {
      finalChildren.unshift(iconComponent)
    } else {
      finalChildren.push(iconComponent)
    }
  }

  const variantType = getVariantType(variant!)
  const variantClassName = disabled
    ? classes[`${variantType}Disabled`]
    : classes[kebabToCamelCase(variant!)]
  const sizeClassName = classes[size!]

  const rootClassName = cx(
    {
      [classes.fullWidth]: fullWidth,
      [classes.active]: active,
      [classes.focused]: focused,
      [classes.hovered]: hovered,
      [classes.circular]: circular
    },
    sizeClassName,
    variantClassName,
    rootClass
  )

  return (
    <ButtonBase
      // eslint-disable-next-line react/jsx-props-no-spreading
      {...rest}
      ref={ref}
      classes={{
        root: rootClassName
      }}
      onClick={onClick}
      className={className}
      style={style}
      disabled={disabled}
      title={title}
      value={value}
      type={type}
    >
      <Container
        as='span'
        inline
        flex
        direction='row'
        alignItems='center'
        className={cx({ [hiddenClass]: loading }, contentClass)}
      >
        {finalChildren}
      </Container>

      {loading && (
        <Loader variant='inherit' className={loaderClass} inline size='small' />
      )}
    </ButtonBase>
  )
}) as CompoundedComponentWithRef<Props, HTMLButtonElement, StaticProps>

Button.defaultProps = {
  active: false,
  children: null,
  circular: false,
  disabled: false,
  focused: false,
  fullWidth: false,
  hovered: false,
  iconPosition: 'left',
  loading: false,
  onClick: () => {},
  size: 'medium',
  type: 'button',
  variant: 'primary-blue'
}

Button.displayName = 'Button'

Button.Group = Group

export default withStyles(styles)(Button) as PicassoComponentWithRef<
  Props,
  HTMLButtonElement,
  StaticProps
>
