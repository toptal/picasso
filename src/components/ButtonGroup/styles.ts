import { createStyles } from '@material-ui/core/styles'

export default () =>
  createStyles({
    root: {
      display: 'flex',
      justifyContent: 'flex-start',

      '& $button + $button': {
        marginLeft: '-1px'
      }
    },
    button: {
      transitionProperty: 'color, background',

      '&:first-child': {
        borderTopRightRadius: 0,
        borderBottomRightRadius: 0
      },

      '&:not(:first-child):not(:last-child)': {
        borderRadius: 0
      },

      '&:last-child': {
        borderTopLeftRadius: 0,
        borderBottomLeftRadius: 0
      },

      '&:active, &$active, &:hover': {
        zIndex: 1
      }
    },
    active: {}
  })
