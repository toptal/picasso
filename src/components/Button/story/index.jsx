import PicassoBook from '~/.storybook/components/PicassoBook'

import buttonGroupStory from '@components/ButtonGroup/story'

import { Button } from '../Button'

const page = PicassoBook.createPage(
  'Button',
  'A Button indicates a possible user action.'
)

page
  .createTabChapter('Props')
  .addComponentDocs({
    component: Button,
    additionalDocs: {
      icon: {
        type: 'ReactElement'
      }
    },
    name: 'Button'
  })
  .addComponentDocs(buttonGroupStory.componentDocs)

page
  .createChapter()
  .addExample('Button/story/Basic.example.jsx', 'Basic')
  .addExample('Button/story/Variants.example.jsx', 'Variants')
  .addExample('Button/story/States.example.jsx', 'States')
  .addExample('Button/story/Disabled.example.jsx', {
    title: 'Disabled',
    description: 'The button shows that currently unable to be interacted with'
  })
  .addExample('Button/story/Sizes.example.jsx', 'Sizes')
  .addExample('Button/story/FullWidth.example.jsx', 'Full width')
  .addExample('Button/story/IconButtons.example.jsx', 'Button with Icon')
  .addExample(
    'Button/story/CircularIconButton.example.jsx',
    'Circular Style Button'
  )
  .addExample(
    'Button/story/IconButtonsWithText.example.jsx',
    'Button with text and Icon'
  )
  .addExample('Button/story/Loading.example.jsx', 'Button with loading state')

page.connect(buttonGroupStory.chapter)
