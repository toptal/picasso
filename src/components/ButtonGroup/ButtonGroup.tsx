import React, { ReactNode, HTMLAttributes, forwardRef } from 'react'
import { withStyles } from '@material-ui/core/styles'
import cx from 'classnames'

import Button from '../Button'
import { StandardProps } from '../Picasso'
import { withClasses } from '../styles'
import styles from './styles'

interface Props extends StandardProps, HTMLAttributes<HTMLDivElement> {
  /** List of `Button` components which you want to render as `ButtonGroup` */
  children?: ReactNode
}

export const ButtonGroup = forwardRef<HTMLDivElement, Props>(
  function ButtonGroup({ children, classes, className, style, ...rest }, ref) {
    return (
      <div
        // eslint-disable-next-line react/jsx-props-no-spreading
        {...rest}
        ref={ref}
        className={cx(classes.root, className)}
        style={style}
      >
        {children}
      </div>
    )
  }
)

ButtonGroup.defaultProps = {
  classes: {}
}

ButtonGroup.displayName = 'ButtonGroup'

export default withStyles(styles)(
  withClasses(classes => [
    [
      Button,
      {
        root: classes.button,
        active: classes.active
      }
    ]
  ])(ButtonGroup)
)
