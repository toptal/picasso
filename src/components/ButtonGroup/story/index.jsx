import PicassoBook from '~/.storybook/components/PicassoBook'

import { ButtonGroup } from '../ButtonGroup'

const chapter = PicassoBook.connectToPage(page =>
  page
    .createChapter(
      'Group of buttons',
      'You can combine multiple buttons into a single container.'
    )
    .addExample('ButtonGroup/story/ButtonGroup.example.jsx', 'Button group')
)

const componentDocs = PicassoBook.createComponentDocs(
  ButtonGroup,
  'Button.Group'
)

export default {
  chapter,
  componentDocs
}
