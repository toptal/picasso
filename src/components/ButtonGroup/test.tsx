import React, { ReactNode } from 'react'
import { render, cleanup } from '@testing-library/react'

import ButtonGroup from './ButtonGroup'
import Button from '../Button'
import Picasso from '../Picasso/Picasso'

const renderButtonGroup = (children: ReactNode) => {
  return render(
    <Picasso loadFonts={false}>
      <ButtonGroup>{children}</ButtonGroup>
    </Picasso>
  )
}

afterEach(cleanup)

describe('ButtonGroup', () => {
  test('render', () => {
    const { container } = renderButtonGroup([
      <Button key='1' />,
      <Button key='2' active />,
      <Button key='3' />
    ])

    expect(container).toMatchSnapshot()
  })
})
