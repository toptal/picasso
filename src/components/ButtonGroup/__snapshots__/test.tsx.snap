// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ButtonGroup render 1`] = `
<div>
  <div
    class="Picasso-root"
  >
    <div
      class="ButtonGroup-root"
    >
      <button
        class="MuiButtonBase-root Button-medium Button-primaryBlue Button-root ButtonGroup-button"
        tabindex="0"
        type="button"
      >
        <span
          class="Component-flex Component-inline Button-content"
          style="flex-direction: row; align-items: center;"
        />
      </button>
      <button
        class="MuiButtonBase-root Button-active ButtonGroup-active Button-medium Button-primaryBlue Button-root ButtonGroup-button"
        tabindex="0"
        type="button"
      >
        <span
          class="Component-flex Component-inline Button-content"
          style="flex-direction: row; align-items: center;"
        />
      </button>
      <button
        class="MuiButtonBase-root Button-medium Button-primaryBlue Button-root ButtonGroup-button"
        tabindex="0"
        type="button"
      >
        <span
          class="Component-flex Component-inline Button-content"
          style="flex-direction: row; align-items: center;"
        />
      </button>
    </div>
  </div>
</div>
`;
